storage/storage_image.go

```diff
@@ -0,0 +1,430 @@
+	if err != nil && err != io.EOF {
+		return "", -1, err
+	}
+	// Set up to read the whole blob (the initial snippet, plus the rest)
+	// while digesting it with sha256.
+	hasher := sha256.New()
+	defragmented := io.MultiReader(bytes.NewBuffer(header[:n]), stream)
+	multi := io.TeeReader(defragmented, hasher)
+	if (n > 0) && archive.IsArchive(header[:n]) {
+		// It's a filesystem layer.  If it's not the first one in the
+		// image, we assume that the most recently added layer is its
+		// parent.
+		parentLayer := ""
+		if len(s.BlobList) > 0 {
+			for _, blob := range s.BlobList {
+				if layerID, ok := s.Layers[blob]; ok {
+					parentLayer = layerID
+				}
+			}
+		}
+		// Now try to figure out if the identifier we have is an ID or
+		// something else, so that we can do collision detection
+		// correctly.
+		names := []string{digest}
+		id := digest
+		if matches := idRegexp.FindStringSubmatch(digest); len(matches) > 1 {
+			id = matches[len(matches)-1]
+			names = nil
+		}
+		layer, err := s.store.CreateLayer(id, parentLayer, names, "", true)
+		if err != nil && err != storage.ErrDuplicateID {
+			return "", -1, err
+		}
+		if err == storage.ErrDuplicateID {
+			// There's already a layer with the same ID.  If its
+			// checksum is the same as ours, then we're cool with
+			// just using the already-stored copy.
+			io.Copy(ioutil.Discard, multi)
+			hash := hasher.Sum(nil)
+			hasher = sha256.New()
+			diff, _, err := s.diffLayer(id, false)
+			if err != nil {
+				logrus.Debugf("blob %q already present but got %v while verifying it", err)
+				return "", -1, err
+			}
+			io.Copy(hasher, diff)
+			diff.Close()
+			if bytes.Compare(hash, hasher.Sum(nil)) != 0 {
+				// They are not the same.  Return an error.
+				logrus.Debugf("blob %q already present with different contents", id)
+				return "", -1, storage.ErrDuplicateID
+			}
+			// They are the same.  Record that this blob is a layer.
+			logrus.Debugf("blob %q already present with same contents, continuing", id)
+			if layer, err = s.store.GetLayer(id); err == nil && layer != nil {
+				id = layer.ID
+			}
+			s.Layers[digest] = id
+		} else {
+			// Apply the layer.
+			actualSize, err = s.store.ApplyDiff(layer.ID, multi)
+			if err != nil {
+				s.store.DeleteLayer(layer.ID)
+				return "", -1, err
+			}
+			// If the diff was compressed, make a note of how it
+			// was compressed, so that we can compress it again if
+			// we need to extract it via GetBlob().
+			layerdata := storageLayerMetadata{
+				Compression:  archive.DetectCompression(header[:n]),
+				ExpectedSize: expectedSize,
+			}
+			if metadata, err := json.Marshal(layerdata); len(metadata) != 0 && err == nil {
+				s.store.SetMetadata(layer.ID, string(metadata))
+			}
+			// Record that this blob is a layer.
+			s.Layers[digest] = layer.ID
+			logrus.Debugf("blob %q imported as a filesystem layer", id)
+		}
+		// Check that our computed sha256sum matches the digest, then
+		// add it to the blob list.
```

Comment:
PutBlob() is going to have to do verification when it can, so that it can know when the incoming data stream turns out to be not what was promised, and that it needs to clean up the layer.  Otherwise a verification failure flagged only at the copy.Image() level is going to end up leaving the layer (with contents that don't match the indicated ID) in the storage library, and it's not easy to recover from that.